onnxruntime/core/framework/session_state.h

```diff
@@ -59,6 +61,30 @@ struct MemoryPatternGroup;
 class MemoryInfo;
 #endif
 
+#ifdef ENABLE_TRAINING
+struct PartialGraphExecutionState {
+ public:
+  PartialGraphExecutionState() {
+    execution_frame_ = nullptr;
+  }
+
+  ~PartialGraphExecutionState() = default;
+
+  void SetProgramCounterStart(size_t start) { program_counter_start_ = start; }
+  void SetProgramCounterEnd(size_t end) { program_counter_end_ = end; }
+
+  size_t GetProgramCounterStart() { return program_counter_start_; }
+  size_t GetProgramCounterEnd() { return program_counter_end_; }
+
+  std::unique_ptr<ExecutionFrame>& GetExecutionFrame() { return execution_frame_; }
```

Comment:
>std::unique_ptr<ExecutionFrame>& [](http://example.com/codeflow?start=2&length=32)

returning a non-const ref of class member is not a good practice IMO. I'd suggest separate it into Get/Set to be more explicit if you really need to change the member, or just return const ref of the class member with a const Get function here.

Reply:
I agree. You're losing control of a key internal by doing this. Add a setter to explicitly transfer ownership, and the getter should return [const] ExecutionFrame& for usage (i.e. the unique_ptr stays internal so the ownership semantics are clear).

Reply:
Is it possible to construct this `ExecutionFrame` when the `PartialGraphExecutionState` is constructed, and get rid of the `unique_ptr` indirection (and uninitialized state) altogether? 

(This would presumably need a bit of refactoring on the Python wrapper for constructing it - maybe you'd need to make it a method on `TrainingAgent` rather than a freestanding constructor? - but I think it should be possible.)

Reply:
@mrry That would be great but it also means we will be moving a fair chunk of code out of ExecutePartialGraph here since we need to do a device check and perform any copies before ExecutionFrame is constructed. It is possible but finding it hard to justify the ow to wow ratio. Let me think a little more on this but please continue to review, I will try this out and get back.

Reply:
@mrry let me think a little more on this, I have addressed Scott's and Ke's comments in the meantime since they were more pressing.

Reply:
I don't see the updated code in the PR - it's still returning `std::unique_ptr<ExecutionFrame>&`. Can you ping me when it's pushed?

Reply:
Spoke offline with @mrry and decided that its ok to keep a setter and getter for `std::unique_ptr<ExecutionFrame>&. ` for now since his suggestion would be disruptive for now.